File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

bool word_check(string a, string b) {
    bool same = true;
    for (int i = 0; i < a.size(); ++i) {
        if (a[i] != b[i]) {
            if (!same) return false;
            same = false;
        }
    }
    return !same;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    vector<int> visited(words.size(), 0);
    queue<pair<string, int>> q;
    q.push({begin, 0});
    while (!q.empty()) {
        string word = q.front().first;
        int count = q.front().second;
        q.pop();
        for (int i = 0; i < words.size(); ++i) {
            if (word_check(word, words[i])) {
                visited[i] = true;
                if (target == words[i]) return count + 1;
                q.push({words[i], count + 1});
            }
        }
    }
    return answer;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.kt
import java.util.LinkedList
import java.util.Queue

class Solution {

    private lateinit var visited: BooleanArray

    fun solution(begin: String, target: String, words: Array<String>): Int {
        var answer = 0
        visited = BooleanArray(words.size) { false }

        val q: Queue<Pair<String, Int>> = LinkedList()
        q.add(Pair(begin, 0))
        while (q.isNotEmpty()) {
            val cur = q.poll()
            val word = cur.first
            val count = cur.second
            if (word == target) return count
            for (i in words.indices) {
                if (!visited[i] && checkWord(words[i], word)){
                    q.add(Pair(words[i], count + 1))
                    visited[i] = true
                }
            }
        }

        return answer
    }

    private fun checkWord(w1: String, w2: String): Boolean {
        var same = true
        for (i in w1.indices) {
            if(w1[i] != w2[i]){
                if (!same) return false
                same = false
            }
        }
        return !same
    }
}